src/openapi/streaming/streaming.ts

```diff
@@ -0,0 +1,1124 @@
+     * Sends an update to a subscription by finding it and calling its callback
+     * @param update - update
+     */
+    private sendDataUpdateToSubscribers(update: types.StreamingMessage) {
+        const subscription = this.findSubscriptionByReferenceId(
+            update.ReferenceId,
+        );
+        if (!subscription || subscription.onStreamingData(update) === false) {
+            // happens if we've been sent to another server and cannot kill the old subscription
+            log.debug(
+                LOG_AREA,
+                'Data update does not match a subscription',
+                update,
+            );
+        }
+    }
+
+    private getHeartbeats(message: types.HeartbeatsControlMessage) {
+        if (message.Heartbeats) {
+            return message.Heartbeats;
+        }
+
+        if (message.Data && message.Data.length > 0) {
+            return message.Data[0].Heartbeats;
+        }
+
+        return null;
+    }
+
+    private getTargetReferenceIds(message: types.ResetControlMessage) {
+        if (message.TargetReferenceIds) {
+            return message.TargetReferenceIds;
+        }
+
+        if (message.Data && message.Data.length > 0) {
+            return message.Data[0].TargetReferenceIds;
+        }
+
+        return null;
+    }
+
+    /**
+     * Handles a control message on the streaming connection
+     * @param message - message from open-api
+     */
+    private handleControlMessage(message: types.ControlMessage) {
+        switch (message.ReferenceId) {
+            case OPENAPI_CONTROL_MESSAGE_HEARTBEAT:
+                this.handleControlMessageFireHeartbeats(
+                    this.getHeartbeats(message),
+                );
+                break;
+
+            case OPENAPI_CONTROL_MESSAGE_RESET_SUBSCRIPTIONS:
+                this.handleControlMessageResetSubscriptions(
+                    this.getTargetReferenceIds(message),
+                );
+                break;
+
+            case OPENAPI_CONTROL_MESSAGE_RECONNECT:
+                this.handleControlMessageReconnect();
+                break;
+
+            case OPENAPI_CONTROL_MESSAGE_DISCONNECT:
+                this.handleControlMessageDisconnect();
+                break;
+
+            default:
+                log.warn(LOG_AREA, 'Unrecognised control message', {
+                    message,
+                    transport: this.getActiveTransportName(),
+                });
+                break;
+        }
+    }
+
+    /**
+     * Fires heartbeats to relevant subscriptions
+     * @param heartbeatList - heartbeatList
+     */
+    private handleControlMessageFireHeartbeats(
+        heartbeatList: types.Heartbeats[] | null,
+    ) {
+        log.debug(LOG_AREA, 'heartbeats received', { heartbeatList });
+        // @ts-expect-error FIXME heartbeatList may be null, decide whether to throw or use an empty array as a default
+        for (let i = 0; i < heartbeatList.length; i++) {
+            // @ts-expect-error

```

Comment:
why expect an error?

Reply:
connection with the suppression above, fixed